python/pyspark/sql/dataframe.py

```diff
@@ -2188,6 +2188,150 @@ def cube(self, *cols: "ColumnOrName") -> "GroupedData":  # type: ignore[misc]
+        by `variableColumnName` and `valueColumnName`.
+
+        When no "id" columns are given, the unpivoted DataFrame consists of only the
+        "variable" and "value" columns.
+
+        All "value" columns must share a least common data type. Unless they are the same data type,
+        all "value" columns are cast to the nearest common data type. For instance, types
+        `IntegerType` and `LongType` are cast to `LongType`, while `IntegerType` and `StringType`
+        do not have a common data type and `unpivot` fails.
+
+        :func:`groupby` is an alias for :func:`groupBy`.
+
+        .. versionadded:: 3.4.0
+
+        Parameters
+        ----------
+        ids : str, Column, tuple, list, optional
+            Column(s) to use as identifiers. Can be a single column or column name,
+            or a list for multiple columns.
+        values : str, Column, tuple, list, optional
+            Column(s) to unpivot. Can be a single column or column name, or a list
+            for multiple columns. If not specified or empty, uses all columns that
+            are not set as `ids`.
+        variableColumnName : scalar, default 'variable'
+            Name of the variable column. If None it uses 'variable'.
+        valueColumnName : scalar, default 'value'
+            Name of the value column. If None it uses 'value'.
+
+        Returns
+        -------
+        DataFrame
+            Unpivoted DataFrame.
+
+        Examples
+        --------
+        >>> df = spark.createDataFrame(
+        ...     [(1, 11, 1.1), (2, 12, 1.2)],
+        ...     ["id", "int", "double"],
+        ... )
+        >>> df.show()
+        +---+---+------+
+        | id|int|double|
+        +---+---+------+
+        |  1| 11|   1.1|
+        |  2| 12|   1.2|
+        +---+---+------+
+
+        >>> df.unpivot("id").show()
+        +---+--------+-----+
+        | id|variable|value|
+        +---+--------+-----+
+        |  1|     int| 11.0|
+        |  1|  double|  1.1|
+        |  2|     int| 12.0|
+        |  2|  double|  1.2|
+        +---+--------+-----+
+        """
+        if ids is None:
+            id_list = []
+        elif isinstance(ids, tuple):
+            id_list = list(ids)
+        elif isinstance(ids, list):
+            id_list = ids
+        else:
+            id_list = [ids]
+
+        if values is None:
+            value_list = []
+        elif isinstance(values, tuple):
+            value_list = list(values)
+        elif isinstance(values, list):
+            value_list = values
+        else:
+            value_list = [values]
+
+        if variableColumnName is None:
+            variableColumnName = "variable"
+
+        if valueColumnName is None:
+            valueColumnName = "value"
+
+        return DataFrame(
+            self._jdf._unpivot(
+                self._jcols(*id_list), self._jcols(*value_list), variableColumnName, valueColumnName
+            ),
+            self.sparkSession,
+        )
+
+    def melt(
```

Comment:
there was also a `melt` function in `python/pyspark/pandas/frame.py`, is it possbile to apply the new `unpivot` implementation there?

Reply:
That is planned, but I'd prefer to do that in a separate PR: SPARK-39878